plaid/src/apis/github/environments.rs

```diff
@@ -0,0 +1,213 @@
+use std::collections::HashMap;
+
+use serde::Serialize;
+use serde_json::Value;
+
+use alkali::asymmetric::seal;
+
+use crate::apis::{github::GitHubError, ApiError};
+
+use super::Github;
+
+#[derive(Serialize)]
+struct DeploymentBranchPolicy {
+    protected_branches: bool,
+    custom_branch_policies: bool,
+}
+
+#[derive(Serialize)]
+struct CreateEnvironmentPayload {
+    wait_timer: u16,
+    prevent_self_review: bool,
+    reviewers: Vec<String>, // This is not true (items are not strings) but we will leave it empty, so we don't care
+    deployment_branch_policy: DeploymentBranchPolicy,
+}
+
+#[derive(Serialize)]
+struct CreateDeploymentBranchPolicyPayload {
+    name: String,
+    #[serde(rename = "type")]
+    type_: String,
+}
+
+#[derive(Serialize)]
+struct UploadEnvironmentSecretPayload {
+    encrypted_value: String,
+    key_id: String,
+}
+
+impl Github {
+    /// Create a new GitHub deployment environment for a given repository
+    /// See https://docs.github.com/en/rest/deployments/environments?apiVersion=2022-11-28#create-or-update-an-environment for more detail
+    pub async fn create_environment_for_repo(
+        &self,
+        params: &str,
+        module: &str,
+    ) -> Result<u32, ApiError> {
+        const ENVIRONMENT_NAME: &str = "publish";
+
+        let request: HashMap<&str, &str> =
+            serde_json::from_str(params).map_err(|_| ApiError::BadRequest)?;
+
+        let owner = self.validate_username(request.get("owner").ok_or(ApiError::BadRequest)?)?;
+        let repo =
+            self.validate_repository_name(request.get("repo").ok_or(ApiError::BadRequest)?)?;
+        let branch =
+            self.validate_branch_name(request.get("branch").ok_or(ApiError::BadRequest)?)?;
+
+        info!(
+            "Creating and configuring 'publish' environment in repo [{repo}] owned by [{owner}] on behalf of [{module}]"
+        );
+
+        // 1. Create the environment
+
+        let address = format!("/repos/{owner}/{repo}/environments/{ENVIRONMENT_NAME}");
+
+        let body = CreateEnvironmentPayload {
+            wait_timer: 0,
+            prevent_self_review: false,
+            reviewers: vec![],
+            deployment_branch_policy: DeploymentBranchPolicy {
+                protected_branches: false,
+                custom_branch_policies: true,
```

Comment:
To be checked if this is exactly what we want, if we should make this configurable, etc.

Reply:
I would ask Andrew

Reply:
Andrew has been triggered 😄